extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+                                  ).collect(Collectors.toList());
+
+                final List<Object[]> retVal = new ArrayList<>();
+                while (!cursor.isDone()) {
+                  final Object[] row = new Object[columnMappings.getMappings().size()];
+                  for (int i = 0; i < row.length; i++) {
+                    row[i] = selectors.get(i).getObject();
+                  }
+                  retVal.add(row);
+                  cursor.advance();
+                }
+
+                return Sequences.simple(retVal);
+              }
+          ).withBaggage(resultReaderExec::shutdownNow)
+      );
+    } else {
+      return null;
+    }
+  }
+
+  private void publishSegmentsIfNeeded(
+      final QueryDefinition queryDef,
+      final ControllerQueryKernel queryKernel
+  ) throws IOException
+  {
+    if (queryKernel.isSuccess() && MSQControllerTask.isIngestion(task.getQuerySpec())) {
+      final StageId finalStageId = queryKernel.getStageId(queryDef.getFinalStageDefinition().getStageNumber());
+
+      //noinspection unchecked
+      @SuppressWarnings("unchecked")
+      final Set<DataSegment> segments = (Set<DataSegment>) queryKernel.getResultObjectForStage(finalStageId);
+      log.info("Query [%s] publishing %d segments.", queryDef.getQueryId(), segments.size());
+      publishAllSegments(segments);
+    }
+  }
+
+  private void cleanUpDurableStorageIfNeeded()
+  {
+    if (MultiStageQueryContext.isDurableStorageEnabled(task.getQuerySpec().getQuery().getContext())) {
+      final String controllerDirName = StringUtils.format("controller_%s", task.getId());
+      try {
+        // Delete all temporary files as a failsafe
+        MSQTasks.makeStorageConnector(context.injector()).deleteRecursively(controllerDirName);
+      }
+      catch (Exception e) {
+        // If an error is thrown while cleaning up a file, log it and try to continue with the cleanup
+        log.warn(e, "Error while cleaning up temporary files at path %s", controllerDirName);
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static QueryDefinition makeQueryDefinition(
+      final String queryId,
+      @SuppressWarnings("rawtypes") final QueryKit toolKit,
+      final MSQSpec querySpec
+  )
+  {
+    final MSQTuningConfig tuningConfig = querySpec.getTuningConfig();
+    final ShuffleSpecFactory shuffleSpecFactory;
+
+    if (MSQControllerTask.isIngestion(querySpec)) {
+      shuffleSpecFactory = (clusterBy, aggregate) ->
+          new TargetSizeShuffleSpec(
+              clusterBy,
+              tuningConfig.getRowsPerSegment(),
+              aggregate
+          );
+    } else if (querySpec.getDestination() instanceof TaskReportMSQDestination) {
+      shuffleSpecFactory = ShuffleSpecFactories.singlePartition();
+    } else {
+      throw new ISE("Unsupported destination [%s]", querySpec.getDestination());
+    }
+
+    final QueryDefinition queryDef;
+
+    try {
+      //noinspection unchecked
```

Comment:
Can we use a usual Java annotation so that all IDEs know to not complain about the unchecked issue?

More generally, can the work be type-safe?

Reply:
I removed the comment and my IDE doesn't complain at me. I guess it wasn't necessary.